src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ValueTaskSource.cs

```diff
@@ -173,5 +187,34 @@ void IValueTaskSource.OnCompleted(Action<object?> continuation, object? state, s
         => _valueTaskSource.OnCompleted(continuation, state, token, flags);
 
     void IValueTaskSource.GetResult(short token)
-        => _valueTaskSource.GetResult(token);
+    {
+        try
+        {
+            _valueTaskSource.GetResult(token);
+        }
+        finally
+        {
+            lock (this)
+            {
+                State state = _state;
+
+                // In case of a cancellation, reset the task and set the stored results if necessary.
+                if (_valueTaskSource.GetStatus(_valueTaskSource.Version) == ValueTaskSourceStatus.Canceled)
+                {
+                    _valueTaskSource.Reset();
+                    if (state == State.Completed)
+                    {
+                        if (_exception is not null)
+                        {
+                            _valueTaskSource.SetException(_exception);
+                        }
+                        else
+                        {
+                            _valueTaskSource.SetResult(true);
```

Comment:
this all is soooo confusing now 😖😵‍💫😅

is this ValueTaskSource only ever used for closing the connection? or is it used for streams as well?

Reading/writing I believe use ResettableValueTaskSource which is _completely_ unrelated to this one, right?

Reply:
Changes in VTS reverted.